hpms/src/main/java/gov/cms/ab2d/hpms/service/HPMSAuthServiceImpl.java

```diff
@@ -0,0 +1,93 @@
+package gov.cms.ab2d.hpms.service;
+
+import gov.cms.ab2d.hpms.hmsapi.HPMSAuthResponse;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Service;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Flux;
+
+import java.time.Duration;
+
+import static org.springframework.http.HttpHeaders.*;
+
+@Service
+public class HPMSAuthServiceImpl implements HPMSAuthService {
+
+    @Value("${hpms.auth.url}")
+    private String authURL;
+
+    @Value("${HPMS_AUTH_KEY_ID}")
+    private String hpmsAPIKeyId;
+
+    @Value("${HPMS_AUTH_KEY_SECRET}")
+    private String hpmsSecret;
+
+    private volatile String authToken;
+
+    private volatile long tokenExpires;
+
+    @Override
+    public void buildAuthHeaders(HttpHeaders headers) {
+        headers.set("X-API-CONSUMER-ID", hpmsAPIKeyId);
+        headers.set(AUTHORIZATION, retrieveAuthToken());
+    }
+
+    private String retrieveAuthToken() {
+        final long currentTimestamp = System.currentTimeMillis();
+        if (authToken == null || currentTimestamp >= tokenExpires) {
+            refreshToken(currentTimestamp);
+        }
+
+        return authToken;
+    }
+
+    private void refreshToken(long currentTimestamp) {
+        authToken = null;
+
+        Flux<HPMSAuthResponse> orgInfoFlux = WebClient.create(authURL)
+                .post()
+                .contentType(MediaType.APPLICATION_JSON)
+                .bodyValue(retrieveAuthRequestPayload())
+                .accept(MediaType.APPLICATION_JSON)
+                .retrieve()
+                .bodyToFlux(HPMSAuthResponse.class);
+
+        // Cough up blood if we can't get an Auth response in a minute.
+        HPMSAuthResponse authResponse = orgInfoFlux.blockFirst(Duration.ofMinutes(1));
+        if (authResponse == null) {
+            throw new RuntimeException("Failed to procure Auth Token");
+        }
+
+        // Convert seconds to millis at a 90% level to pad refreshing of a token so that we are not in the middle of
+        // a significant operation when the token expires.
+        tokenExpires = currentTimestamp + authResponse.getExpires() * 900;
+        authToken = authResponse.getAccessToken();
+    }
+
+    private String retrieveAuthRequestPayload() {
+        return String.format(AUTH_PAYLOAD_TEMPLATE, hpmsAPIKeyId, hpmsSecret);
+    }
+
+    // Text Blocks are still a preview feature in JDK 13 and 14, otherwise, it would really clean this up.
+    private static final String AUTH_PAYLOAD_TEMPLATE = "{\n" +
```

Comment:
It's a little more typesafe to do these as actual java objects and have them translated by an ObjectMapper or GSON but this works.

Reply:
If it was more complicated, then I would do something more sophisticated.  I just try not to pull in more libraries/objects unless their weight justifies their use.  For this case, I don't really care about all the other stuff, just need to insert secrets to get the token.